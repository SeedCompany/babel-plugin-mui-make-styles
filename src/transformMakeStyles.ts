import { NodePath } from '@babel/traverse';
import {
  CallExpression,
  ImportSpecifier,
  isIdentifier,
  isImportDeclaration,
  isProperty,
  ObjectExpression,
  Program,
} from '@babel/types';
import { basename, extname } from 'path';
import { BabelPlugin } from './util';

interface State {
  // List of imported names of makeStyles.
  // Should typically just be `['makeStyles']`
  names: string[];
}

/**
 * A plugin for Material UI's makeStyles.
 *
 * This sets the classNamePrefix option based on the filename if it has not
 * been set. Which means instead of all the css classes being prefixed with
 * "makeStyles-" we see the component name (by convention).
 */
export const transformMakeStyles: BabelPlugin<State> = ({ types }) => ({
  name: 'mui-styles',
  visitor: {
    Program: (path: NodePath<Program>, state) => {
      // Reset list on new file
      state.names = [];
    },
    ImportSpecifier: (path: NodePath<ImportSpecifier>, state) => {
      // Add name to list if it matches
      if (
        path.node.imported.name === 'makeStyles' &&
        isImportDeclaration(path.parent) &&
        path.parent.source.value.startsWith('@material-ui/')
      ) {
        state.names.push(path.node.local.name);
      }
    },
    CallExpression: (path: NodePath<CallExpression>, state) => {
      // Continue if we are a call to one of our makeStyle functions
      if (
        !isIdentifier(path.node.callee) ||
        !state.names.includes(path.node.callee.name)
      ) {
        return;
      }

      // Find options argument or create new it
      const options =
        (path.node.arguments[1] as ObjectExpression) ||
        types.objectExpression([]);

      // Continue if options doesn't have a classNamePrefix
      if (options.properties.some(isPropertyName('classNamePrefix'))) {
        return;
      }

      // Determine prefix based on filename
      const filePath: string = path.hub.file.opts.filename;
      const classNamePrefix = basename(filePath, extname(filePath));

      // Add it to the options object
      options.properties.unshift(
        types.objectProperty(
          types.identifier('classNamePrefix'),
          types.stringLiteral(classNamePrefix)
        )
      );

      // Replace function call with same one with options argument added.
      path.replaceWith(
        types.callExpression(path.node.callee, [
          path.node.arguments[0],
          options,
        ])
      );

      // Don't traverse into newly replaced path (infinite loop)
      path.skip();
    },
  },
});

const isPropertyName = (name: string) => (prop: { type: unknown }) =>
  isProperty(prop) && isIdentifier(prop.key) && prop.key.name === name;
