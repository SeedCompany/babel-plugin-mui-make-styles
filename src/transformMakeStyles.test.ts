import { transform } from '@babel/core';
import { transformMakeStyles } from './transformMakeStyles';

// language=TypeScript
const cases = {
  basic: `
    import { makeStyles } from '@material-ui/core';
    const useStyles = makeStyles({
      foo: { color: 'red' }
    });
  `,
  'imported from core/styles': `
    import { makeStyles } from '@material-ui/core/styles';
    const useStyles = makeStyles({
      foo: { color: 'red' }
    });
  `,
  'imported from core/styles/makeStyles': `
    import { makeStyles } from '@material-ui/core/styles/makeStyles';
    const useStyles = makeStyles({
      foo: { color: 'red' }
    });
  `,
  'imported from styles': `
    import { makeStyles } from '@material-ui/styles';
    const useStyles = makeStyles({
      foo: { color: 'red' }
    });
  `,
  'with renamed import': `
    import { makeStyles as ms } from '@material-ui/core';
    const useStyles = ms({
      foo: { color: 'red' }
    });
  `,
  // Why? Idk, just covering bases
  'with multiple imports': `
    import { makeStyles } from '@material-ui/core/styles';
    import { makeStyles as ms } from '@material-ui/core';
    const useStyles = ms({
      foo: { color: 'red' }
    });
    const useStyles2 = makeStyles({
      foo: { color: 'red' }
    });
  `,
  'with options already defined': `
    import { makeStyles } from '@material-ui/core';
    const useStyles = makeStyles({
      foo: { color: 'red' }
    }, {
      flip: true,
    });
  `,
  'with prefix already defined': `
    import { makeStyles } from '@material-ui/core';
    const useStyles = makeStyles({
      foo: { color: 'red' }
    }, {
      classNamePrefix: 'FooBar',
    });
  `,
};
it.each(Object.entries(cases))(`works %s`, (name, src) => {
  const result = transform(src, {
    plugins: [transformMakeStyles],
    filename: 'MyButton.tsx',
  });
  expect(result?.code).toMatchSnapshot();
});
