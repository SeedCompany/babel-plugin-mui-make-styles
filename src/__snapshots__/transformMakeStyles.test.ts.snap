// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works basic 1`] = `
"import { makeStyles } from '@material-ui/core';
const useStyles = makeStyles({
  foo: {
    color: 'red'
  }
}, {
  classNamePrefix: \\"MyButton\\"
});"
`;

exports[`works imported from core/styles 1`] = `
"import { makeStyles } from '@material-ui/core/styles';
const useStyles = makeStyles({
  foo: {
    color: 'red'
  }
}, {
  classNamePrefix: \\"MyButton\\"
});"
`;

exports[`works imported from core/styles/makeStyles 1`] = `
"import { makeStyles } from '@material-ui/core/styles/makeStyles';
const useStyles = makeStyles({
  foo: {
    color: 'red'
  }
}, {
  classNamePrefix: \\"MyButton\\"
});"
`;

exports[`works imported from styles 1`] = `
"import { makeStyles } from '@material-ui/styles';
const useStyles = makeStyles({
  foo: {
    color: 'red'
  }
}, {
  classNamePrefix: \\"MyButton\\"
});"
`;

exports[`works with multiple imports 1`] = `
"import { makeStyles } from '@material-ui/core/styles';
import { makeStyles as ms } from '@material-ui/core';
const useStyles = ms({
  foo: {
    color: 'red'
  }
}, {
  classNamePrefix: \\"MyButton\\"
});
const useStyles2 = makeStyles({
  foo: {
    color: 'red'
  }
}, {
  classNamePrefix: \\"MyButton\\"
});"
`;

exports[`works with options already defined 1`] = `
"import { makeStyles } from '@material-ui/core';
const useStyles = makeStyles({
  foo: {
    color: 'red'
  }
}, {
  classNamePrefix: \\"MyButton\\",
  flip: true
});"
`;

exports[`works with prefix already defined 1`] = `
"import { makeStyles } from '@material-ui/core';
const useStyles = makeStyles({
  foo: {
    color: 'red'
  }
}, {
  classNamePrefix: 'FooBar'
});"
`;

exports[`works with renamed import 1`] = `
"import { makeStyles as ms } from '@material-ui/core';
const useStyles = ms({
  foo: {
    color: 'red'
  }
}, {
  classNamePrefix: \\"MyButton\\"
});"
`;
